Add DictTensor tensor wrapper subclass

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #146172
* #146101
* #145922
* #141842
* __->__ #141841
* #146052

soulitzer committed:
Add CachedTensor tensor wrapper subclass

[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"




[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"




[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"




[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"




[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"




[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"




[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"




[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"




[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"




[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"




[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"




[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"




[ghstack-poisoned]

jbschlosser commented on torch/nested/_internal/cached_tensor.py:
```diff
@@ -0,0 +1,166 @@
+        metadata: Dict[str, torch.Tensor],
+        source_fields: Tuple[str, ...],
+        extra_fields: Tuple[str, ...],
+        target_field: Optional[str] = None,
```
is `target_field` used in NJT? didn't look like it to me but I could be wrong

soulitzer replied:
The original use case here was if a NJT's metadata holds both the lengths and offsets. If the user calls .offsets(), we'd want to return a metadata with target_field as offsets, and analgously for lengths.

I ended up updating .offsets()/._offsets to just return plain tensors, so this logic is no longer used by NJT today, but I imagine it would be if someone wanted a .offsets_with_cache() API.

jbschlosser reviewed (commented):
Interesting idea to use a tensor subclass here! I think I like it :) I do have some worries about it in its current state though. I find it hard to conceptualize exactly what responsibilities this tensor is handling.

Imo it'd be simpler to reason about to have a tensor subclass that acts like a single tensor but also carries around optional associated metadata. This simplifies the `source_fields`, `target_field`, etc. distinction: there'd be a single tensor this instance acts as and a set of other tensors carted around.

It's more complex and odd to reason about to have the possibility of multiple `source_field`s or tensors that are essentially equivalent and interchangeable (?) I'm not really sure how this is utilized in practice for NJT but it doesn't seem like it could be seamless anyway (correct me if I'm wrong), so might as well remove the complexity.

TL;DR: the responsibility of considering two tensors as equivalent (and the complexity / concepts needed) feels outside of the scope of a cached tensor object and makes it harder to reason about / grok. If it's possible to boot this to a layer above in NJT, I'd prefer that

jbschlosser commented on torch/nested/_internal/cached_tensor.py:
```diff
@@ -0,0 +1,166 @@
+_func_registry: Dict[Any, Callable] = {}
+
+
+@contextmanager
+def set_func_registry(registry: Dict[Any, Callable]) -> Generator:
+    global _func_registry
+    # Save the current registry to restore it later
+    old_registry = _func_registry
+    # Set the new registry
+    _func_registry = registry
+    try:
+        yield
+    finally:
+        # Restore the old registry
+        _func_registry = old_registry
+
+
+# Note: [ CacheTensor open registry ]
+#
+# Registering this decorator for an op allows CachedTensor's torch dispatch to
+# redirect calls to that op to your own implementations.
+# For NestedTensor we rely on this behavior for factory functions.
+def register_cached_tensor_func(aten_op: Any) -> Callable:
+    def wrapper(func: Callable) -> Callable:
+        _func_registry[aten_op] = func
+        return func
+
+    return wrapper
```
I think it's an interesting way to customize the behavior of this class based on what the parent subclass needs. For the lazy: what in particular is this utilized for for NJT / torch.compile?

soulitzer replied:
Specifically, NJT has two constructors from_padded and from_jagged that rely on this. Both require CachedTensor as input.

soulitzer commented on torch/nested/_internal/cached_tensor.py:
```diff
@@ -0,0 +1,166 @@
+            )
+
+    def __tensor_flatten__(self) -> Tuple[List[str], Dict[str, Any]]:
+        ctx = {
+            "source_fields": self.source_fields,
+            "extra_fields": self.extra_fields,
+        }
```
Whoops

soulitzer commented on torch/nested/_internal/cached_tensor.py:
```diff
@@ -0,0 +1,166 @@
+        self.metadata = metadata
+
+        # Compile only
+        self.nested_int_ref: Any = None
```
I guess moving it to weaktensorkeydict on FakeMode from CachedTensor to nested int ref would fully clean up this class

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -8654,6 +8654,111 @@ def test_with_factor(self):
         self.assertTrue(a * 2 == 2 * a)
 
 
+from torch.nested._internal.cached_tensor import CachedTensor
```
I would agree with you haha, sadly it's hard to behave like a true CachedTensor without metadata mutation, so  suggestions welcome/happy to bikeshed some more.

Also given the approach of handling the source + extra fields distinction at a different layer, it could even be something like `ContainerTensor`.

soulitzer commented:
> Interesting idea to use a tensor subclass here! I think I like it :) I do have some worries about it in its current state though. I find it hard to conceptualize exactly what responsibilities this tensor is handling.

> Imo it'd be simpler to reason about to have a tensor subclass that acts like a single tensor but also carries around optional associated metadata. This simplifies the source_fields, target_field, etc. distinction: there'd be a single tensor this instance acts as and a set of other tensors carted around.

> It's more complex and odd to reason about to have the possibility of multiple source_fields or tensors that are essentially equivalent and interchangeable (?) I'm not really sure how this is utilized in practice for NJT but it doesn't seem like it could be seamless anyway (correct me if I'm wrong), so might as well remove the complexity.

> TL;DR: the responsibility of considering two tensors as equivalent (and the complexity / concepts needed) feels outside of the scope of a cached tensor object and makes it harder to reason about / grok. If it's possible to boot this to a layer above in NJT, I'd prefer that

This is great feedback, let me play around with some more :)

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"


A generic CachedTensor tensor wrapper subclass that is not aware of NJTs, nested ints, etc. Users of CachedTensor customize how to use it via the `source_fields`, `extra_fields`, and `target_field`:
- `source_fields` are fields like `offsets` and `lengths` that are complete and transitions between them are information preserving. 
- `extra_fields` on the other hand are not complete. A transition from source_field to extra_field is information losing. You cannot reconstruct a source_field with only an extra_fields. 
- A CachedTensor must contain at least one source_field.
- `target_field` (optional) determines, the metadata of the outer tensor. If no `target_field` is provided, the metadata is determined by that of the first source field.

Read more about how it is used in NJT in the "refactor NJT" PR.


 

[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"


~~A generic CachedTensor tensor wrapper subclass that is not aware of NJTs, nested ints, etc. Users of CachedTensor customize how to use it via the `source_fields`, `extra_fields`, and `target_field`:~~
- ~~`source_fields` are fields like `offsets` and `lengths` that are complete and transitions between them are information preserving. ~~
- ~~`extra_fields` on the other hand are not complete. A transition from source_field to extra_field is information losing. You cannot reconstruct a source_field with only an extra_fields.~~
- ~~A CachedTensor must contain at least one source_field.~~
- ~~`target_field` (optional) determines, the metadata of the outer tensor. If no `target_field` is provided, the metadata is determined by that of the first source field.``~~


A simple ContainerTensor that can hold arbitrary inner tensors (keyed on str). User passes a "source_field" to tell us which of the inner tensors the outer tensor's shape, stride, device etc. should be based off of.

Read more about how it is used in NJT in the "refactor NJT" PR.


 

[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"


~~A generic CachedTensor tensor wrapper subclass that is not aware of NJTs, nested ints, etc. Users of CachedTensor customize how to use it via the `source_fields`, `extra_fields`, and `target_field`:~~
- ~~`source_fields` are fields like `offsets` and `lengths` that are complete and transitions between them are information preserving. ~~
- ~~`extra_fields` on the other hand are not complete. A transition from source_field to extra_field is information losing. You cannot reconstruct a source_field with only an extra_fields.~~
- ~~A CachedTensor must contain at least one source_field.~~
- ~~`target_field` (optional) determines, the metadata of the outer tensor. If no `target_field` is provided, the metadata is determined by that of the first source field.``~~


A simple ContainerTensor that can hold arbitrary inner tensors (keyed on str). User passes a "source_field" to tell us which of the inner tensors the outer tensor's shape, stride, device etc. should be based off of.

Read more about how it is used in NJT in the "refactor NJT" PR.


 

[ghstack-poisoned]

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -8654,6 +8654,111 @@ def test_with_factor(self):
+                )
+
+        # After leaving the context, clone behaves as it did before.
+        cloned_cached_tensor = cached_tensor.clone()
+        self.assertFalse(isinstance(cloned_cached_tensor, CachedTensor))
```
Another example actually where clone does not produce an object with the same type is `nn.Parameter`

```
>>> a = torch.nn.Parameter(torch.tensor(1.))
>>> isinstance(a.clone(), torch.nn.Parameter)
False
```

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"


~~A generic CachedTensor tensor wrapper subclass that is not aware of NJTs, nested ints, etc. Users of CachedTensor customize how to use it via the `source_fields`, `extra_fields`, and `target_field`:~~
- ~~`source_fields` are fields like `offsets` and `lengths` that are complete and transitions between them are information preserving.~~
- ~~`extra_fields` on the other hand are not complete. A transition from source_field to extra_field is information losing. You cannot reconstruct a source_field with only an extra_fields.~~
- ~~A CachedTensor must contain at least one source_field.~~
- ~~`target_field` (optional) determines, the metadata of the outer tensor. If no `target_field` is provided, the metadata is determined by that of the first source field.``~~

Relevant block comment from PR:
```
    # Tensor subclass wrapping a dict of tensors, whose shape, dtype, device, etc.
    # is determined by a "source" tensor in the dict (specified by the user
    # during construction).
    #
    # This class is not super useful on its own because by default, performing any
    # operations on it will be as if you first unwrapped the CachedTensor and then
    # performed the operation on the plain source tensor (a plain tensor is also returned).
    # To leverage the extra metadata, you must register an op to perform the special logic
    # you want via register_cached_tensor_func.
    #
    # When used this way (1) it is a convenient way to keep around metadata
    # related to a tensor, without having to laboriously thread those extra metadata
    # around, e.g. through aten signatures. (2) allows one to trigger custom
    # __torch_dispatch__ logic to construct tensor subclasses (which can be tricky to do
    # otherwise because the subclass constructor op itself usually does not take the
    # subclass itself as input!). See NestedTensor for an example.
    
 ```



 

[ghstack-poisoned]

soulitzer commented:
Ready for review again!

Updates: 
- no longer handling the source/extra distinction at this layer (those are handled at NT level instead)
- instead of holding weakref on the CachedTensor itself, FakeTensorMode holds a WeakTensorKeyDictionary

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"


~~A generic CachedTensor tensor wrapper subclass that is not aware of NJTs, nested ints, etc. Users of CachedTensor customize how to use it via the `source_fields`, `extra_fields`, and `target_field`:~~
- ~~`source_fields` are fields like `offsets` and `lengths` that are complete and transitions between them are information preserving.~~
- ~~`extra_fields` on the other hand are not complete. A transition from source_field to extra_field is information losing. You cannot reconstruct a source_field with only an extra_fields.~~
- ~~A CachedTensor must contain at least one source_field.~~
- ~~`target_field` (optional) determines, the metadata of the outer tensor. If no `target_field` is provided, the metadata is determined by that of the first source field.``~~

Relevant block comment from PR:
```
    # Tensor subclass wrapping a dict of tensors, whose shape, dtype, device, etc.
    # is determined by a "source" tensor in the dict (specified by the user
    # during construction).
    #
    # This class is not super useful on its own because by default, performing any
    # operations on it will be as if you first unwrapped the CachedTensor and then
    # performed the operation on the plain source tensor (a plain tensor is also returned).
    # To leverage the extra metadata, you must register an op to perform the special logic
    # you want via register_cached_tensor_func.
    #
    # When used this way (1) it is a convenient way to keep around metadata
    # related to a tensor, without having to laboriously thread those extra metadata
    # around, e.g. through aten signatures. (2) allows one to trigger custom
    # __torch_dispatch__ logic to construct tensor subclasses (which can be tricky to do
    # otherwise because the subclass constructor op itself usually does not take the
    # subclass itself as input!). See NestedTensor for an example.
    
 ```



 

[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"


~~A generic CachedTensor tensor wrapper subclass that is not aware of NJTs, nested ints, etc. Users of CachedTensor customize how to use it via the `source_fields`, `extra_fields`, and `target_field`:~~
- ~~`source_fields` are fields like `offsets` and `lengths` that are complete and transitions between them are information preserving.~~
- ~~`extra_fields` on the other hand are not complete. A transition from source_field to extra_field is information losing. You cannot reconstruct a source_field with only an extra_fields.~~
- ~~A CachedTensor must contain at least one source_field.~~
- ~~`target_field` (optional) determines, the metadata of the outer tensor. If no `target_field` is provided, the metadata is determined by that of the first source field.``~~

Relevant block comment from PR:
```
    # Tensor subclass wrapping a dict of tensors, whose shape, dtype, device, etc.
    # is determined by a "source" tensor in the dict (specified by the user
    # during construction).
    #
    # This class is not super useful on its own because by default, performing any
    # operations on it will be as if you first unwrapped the CachedTensor and then
    # performed the operation on the plain source tensor (a plain tensor is also returned).
    # To leverage the extra metadata, you must register an op to perform the special logic
    # you want via register_cached_tensor_func.
    #
    # When used this way (1) it is a convenient way to keep around metadata
    # related to a tensor, without having to laboriously thread those extra metadata
    # around, e.g. through aten signatures. (2) allows one to trigger custom
    # __torch_dispatch__ logic to construct tensor subclasses (which can be tricky to do
    # otherwise because the subclass constructor op itself usually does not take the
    # subclass itself as input!). See NestedTensor for an example.
    
 ```



 

[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"


~~A generic CachedTensor tensor wrapper subclass that is not aware of NJTs, nested ints, etc. Users of CachedTensor customize how to use it via the `source_fields`, `extra_fields`, and `target_field`:~~
- ~~`source_fields` are fields like `offsets` and `lengths` that are complete and transitions between them are information preserving.~~
- ~~`extra_fields` on the other hand are not complete. A transition from source_field to extra_field is information losing. You cannot reconstruct a source_field with only an extra_fields.~~
- ~~A CachedTensor must contain at least one source_field.~~
- ~~`target_field` (optional) determines, the metadata of the outer tensor. If no `target_field` is provided, the metadata is determined by that of the first source field.``~~

Relevant block comment from PR:
```
    # Tensor subclass wrapping a dict of tensors, whose shape, dtype, device, etc.
    # is determined by a "source" tensor in the dict (specified by the user
    # during construction).
    #
    # This class is not super useful on its own because by default, performing any
    # operations on it will be as if you first unwrapped the CachedTensor and then
    # performed the operation on the plain source tensor (a plain tensor is also returned).
    # To leverage the extra metadata, you must register an op to perform the special logic
    # you want via register_cached_tensor_func.
    #
    # When used this way (1) it is a convenient way to keep around metadata
    # related to a tensor, without having to laboriously thread those extra metadata
    # around, e.g. through aten signatures. (2) allows one to trigger custom
    # __torch_dispatch__ logic to construct tensor subclasses (which can be tricky to do
    # otherwise because the subclass constructor op itself usually does not take the
    # subclass itself as input!). See NestedTensor for an example.
    
 ```



 

[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"


~~A generic CachedTensor tensor wrapper subclass that is not aware of NJTs, nested ints, etc. Users of CachedTensor customize how to use it via the `source_fields`, `extra_fields`, and `target_field`:~~
- ~~`source_fields` are fields like `offsets` and `lengths` that are complete and transitions between them are information preserving.~~
- ~~`extra_fields` on the other hand are not complete. A transition from source_field to extra_field is information losing. You cannot reconstruct a source_field with only an extra_fields.~~
- ~~A CachedTensor must contain at least one source_field.~~
- ~~`target_field` (optional) determines, the metadata of the outer tensor. If no `target_field` is provided, the metadata is determined by that of the first source field.``~~

Relevant block comment from PR:
```
    # Tensor subclass wrapping a dict of tensors, whose shape, dtype, device, etc.
    # is determined by a "source" tensor in the dict (specified by the user
    # during construction).
    #
    # This class is not super useful on its own because by default, performing any
    # operations on it will be as if you first unwrapped the CachedTensor and then
    # performed the operation on the plain source tensor (a plain tensor is also returned).
    # To leverage the extra metadata, you must register an op to perform the special logic
    # you want via register_cached_tensor_func.
    #
    # When used this way (1) it is a convenient way to keep around metadata
    # related to a tensor, without having to laboriously thread those extra metadata
    # around, e.g. through aten signatures. (2) allows one to trigger custom
    # __torch_dispatch__ logic to construct tensor subclasses (which can be tricky to do
    # otherwise because the subclass constructor op itself usually does not take the
    # subclass itself as input!). See NestedTensor for an example.
    
 ```



 

[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"


~~A generic CachedTensor tensor wrapper subclass that is not aware of NJTs, nested ints, etc. Users of CachedTensor customize how to use it via the `source_fields`, `extra_fields`, and `target_field`:~~
- ~~`source_fields` are fields like `offsets` and `lengths` that are complete and transitions between them are information preserving.~~
- ~~`extra_fields` on the other hand are not complete. A transition from source_field to extra_field is information losing. You cannot reconstruct a source_field with only an extra_fields.~~
- ~~A CachedTensor must contain at least one source_field.~~
- ~~`target_field` (optional) determines, the metadata of the outer tensor. If no `target_field` is provided, the metadata is determined by that of the first source field.``~~

Relevant block comment from PR:
```
    # Tensor subclass wrapping a dict of tensors, whose shape, dtype, device, etc.
    # is determined by a "source" tensor in the dict (specified by the user
    # during construction).
    #
    # This class is not super useful on its own because by default, performing any
    # operations on it will be as if you first unwrapped the CachedTensor and then
    # performed the operation on the plain source tensor (a plain tensor is also returned).
    # To leverage the extra metadata, you must register an op to perform the special logic
    # you want via register_cached_tensor_func.
    #
    # When used this way (1) it is a convenient way to keep around metadata
    # related to a tensor, without having to laboriously thread those extra metadata
    # around, e.g. through aten signatures. (2) allows one to trigger custom
    # __torch_dispatch__ logic to construct tensor subclasses (which can be tricky to do
    # otherwise because the subclass constructor op itself usually does not take the
    # subclass itself as input!). See NestedTensor for an example.
    
 ```



 

[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"


~~A generic CachedTensor tensor wrapper subclass that is not aware of NJTs, nested ints, etc. Users of CachedTensor customize how to use it via the `source_fields`, `extra_fields`, and `target_field`:~~
- ~~`source_fields` are fields like `offsets` and `lengths` that are complete and transitions between them are information preserving.~~
- ~~`extra_fields` on the other hand are not complete. A transition from source_field to extra_field is information losing. You cannot reconstruct a source_field with only an extra_fields.~~
- ~~A CachedTensor must contain at least one source_field.~~
- ~~`target_field` (optional) determines, the metadata of the outer tensor. If no `target_field` is provided, the metadata is determined by that of the first source field.``~~

Relevant block comment from PR:
```
    # Tensor subclass wrapping a dict of tensors, whose shape, dtype, device, etc.
    # is determined by a "source" tensor in the dict (specified by the user
    # during construction).
    #
    # This class is not super useful on its own because by default, performing any
    # operations on it will be as if you first unwrapped the CachedTensor and then
    # performed the operation on the plain source tensor (a plain tensor is also returned).
    # To leverage the extra metadata, you must register an op to perform the special logic
    # you want via register_cached_tensor_func.
    #
    # When used this way (1) it is a convenient way to keep around metadata
    # related to a tensor, without having to laboriously thread those extra metadata
    # around, e.g. through aten signatures. (2) allows one to trigger custom
    # __torch_dispatch__ logic to construct tensor subclasses (which can be tricky to do
    # otherwise because the subclass constructor op itself usually does not take the
    # subclass itself as input!). See NestedTensor for an example.
    
 ```



 

[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"


~~A generic CachedTensor tensor wrapper subclass that is not aware of NJTs, nested ints, etc. Users of CachedTensor customize how to use it via the `source_fields`, `extra_fields`, and `target_field`:~~
- ~~`source_fields` are fields like `offsets` and `lengths` that are complete and transitions between them are information preserving.~~
- ~~`extra_fields` on the other hand are not complete. A transition from source_field to extra_field is information losing. You cannot reconstruct a source_field with only an extra_fields.~~
- ~~A CachedTensor must contain at least one source_field.~~
- ~~`target_field` (optional) determines, the metadata of the outer tensor. If no `target_field` is provided, the metadata is determined by that of the first source field.``~~

Relevant block comment from PR:
```
    # Tensor subclass wrapping a dict of tensors, whose shape, dtype, device, etc.
    # is determined by a "source" tensor in the dict (specified by the user
    # during construction).
    #
    # This class is not super useful on its own because by default, performing any
    # operations on it will be as if you first unwrapped the CachedTensor and then
    # performed the operation on the plain source tensor (a plain tensor is also returned).
    # To leverage the extra metadata, you must register an op to perform the special logic
    # you want via register_cached_tensor_func.
    #
    # When used this way (1) it is a convenient way to keep around metadata
    # related to a tensor, without having to laboriously thread those extra metadata
    # around, e.g. through aten signatures. (2) allows one to trigger custom
    # __torch_dispatch__ logic to construct tensor subclasses (which can be tricky to do
    # otherwise because the subclass constructor op itself usually does not take the
    # subclass itself as input!). See NestedTensor for an example.
    
 ```



 

[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"


~~A generic CachedTensor tensor wrapper subclass that is not aware of NJTs, nested ints, etc. Users of CachedTensor customize how to use it via the `source_fields`, `extra_fields`, and `target_field`:~~
- ~~`source_fields` are fields like `offsets` and `lengths` that are complete and transitions between them are information preserving.~~
- ~~`extra_fields` on the other hand are not complete. A transition from source_field to extra_field is information losing. You cannot reconstruct a source_field with only an extra_fields.~~
- ~~A CachedTensor must contain at least one source_field.~~
- ~~`target_field` (optional) determines, the metadata of the outer tensor. If no `target_field` is provided, the metadata is determined by that of the first source field.``~~

Relevant block comment from PR:
```
    # Tensor subclass wrapping a dict of tensors, whose shape, dtype, device, etc.
    # is determined by a "source" tensor in the dict (specified by the user
    # during construction).
    #
    # This class is not super useful on its own because by default, performing any
    # operations on it will be as if you first unwrapped the CachedTensor and then
    # performed the operation on the plain source tensor (a plain tensor is also returned).
    # To leverage the extra metadata, you must register an op to perform the special logic
    # you want via register_cached_tensor_func.
    #
    # When used this way (1) it is a convenient way to keep around metadata
    # related to a tensor, without having to laboriously thread those extra metadata
    # around, e.g. through aten signatures. (2) allows one to trigger custom
    # __torch_dispatch__ logic to construct tensor subclasses (which can be tricky to do
    # otherwise because the subclass constructor op itself usually does not take the
    # subclass itself as input!). See NestedTensor for an example.
    
 ```



 

[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"


~~A generic CachedTensor tensor wrapper subclass that is not aware of NJTs, nested ints, etc. Users of CachedTensor customize how to use it via the `source_fields`, `extra_fields`, and `target_field`:~~
- ~~`source_fields` are fields like `offsets` and `lengths` that are complete and transitions between them are information preserving.~~
- ~~`extra_fields` on the other hand are not complete. A transition from source_field to extra_field is information losing. You cannot reconstruct a source_field with only an extra_fields.~~
- ~~A CachedTensor must contain at least one source_field.~~
- ~~`target_field` (optional) determines, the metadata of the outer tensor. If no `target_field` is provided, the metadata is determined by that of the first source field.``~~

Relevant block comment from PR:
```
    # Tensor subclass wrapping a dict of tensors, whose shape, dtype, device, etc.
    # is determined by a "source" tensor in the dict (specified by the user
    # during construction).
    #
    # This class is not super useful on its own because by default, performing any
    # operations on it will be as if you first unwrapped the CachedTensor and then
    # performed the operation on the plain source tensor (a plain tensor is also returned).
    # To leverage the extra metadata, you must register an op to perform the special logic
    # you want via register_cached_tensor_func.
    #
    # When used this way (1) it is a convenient way to keep around metadata
    # related to a tensor, without having to laboriously thread those extra metadata
    # around, e.g. through aten signatures. (2) allows one to trigger custom
    # __torch_dispatch__ logic to construct tensor subclasses (which can be tricky to do
    # otherwise because the subclass constructor op itself usually does not take the
    # subclass itself as input!). See NestedTensor for an example.
    
 ```



 

[ghstack-poisoned]

jbschlosser commented on torch/nested/_internal/cached_tensor.py:
```diff
@@ -0,0 +1,131 @@
+    # subclass itself as input!). See NestedTensor for an example.
+
+    @staticmethod
+    @torch._disable_dynamo  # type: ignore[misc]
```
why does this need to be here?

soulitzer replied:
From talking with @bdhirsh its standard practice to prevent dynamo from randomly tracing this frame and erroring out in a weird way. (It's also used in DTensor.)

soulitzer replied:
This reminds me to add a test for constructing CachedTensor in the middle of the graph (which we may rely on for NestedTensor)

jbschlosser commented on torch/nested/_internal/cached_tensor.py:
```diff
@@ -0,0 +1,131 @@
+
+    @staticmethod
+    @torch._disable_dynamo  # type: ignore[misc]
+    def __new__(
+        cls,
+        metadata: Dict[str, Optional[torch.Tensor]],
+        source_field: str,
+    ) -> "CachedTensor":
```
Okay I'm loving how much simpler this class is now! There's still one thing bothering me (from the perspective of reviewing this in isolation without looking much in detail at how it's used for NJT): the concept of a specifying a dict and source field that refers to an object in that dict seems overly restrictive and a bit awkward.

Ideally I'd want to see something like:
```
CacheTensor(t: torch.Tensor, cache: Dict[str, Optional[torch.Tensor]])
```

where the `CacheTensor` operates like `t` but also contains the extra `cache` stuff alongside it, and notably `cache` doesn't contain `t`. This seems about as clean as possible.

Would this cause problems with the way NJT wants to think about `source_field`? It feels like this should still accomplish what NJT needs, albeit slightly differently, but lmk if you think there are unsolvable perf issues with this.

soulitzer replied:
The main difference I see between the two approaches are that one requires the source field to have a name, the other does not. 

For NJT requiring the source field to have a name is useful because it allows me to access every field on the CacheTensor including the source field in a uniform way.

jbschlosser replied:
Hm I'm not super convinced, especially because I want to consider this class outside of an NJT context. The source field thing makes CacheTensor usage marginally simpler within NJT but more awkward for other use cases imo. Treating the source field and some other cached objects uniformly seems pretty specific to the NJT use case, and arguably should be handled at the NJT layer.

> For NJT requiring the source field to have a name is useful because it allows me to access every field on the CacheTensor including the source field in a uniform way.

You can maintain a name within NJT, no? e.g. store canonical source field for a given NJT instance:
```
class NJT:
    ...
    source_field = "_device_offsets"  # for example
    source = CacheTensor(source_tensor, other_tensors)

    # custom getattr returns CacheTensor's t for source_field and indexes other_tensors for other fields.
    # it's less uniform from the perspective of NJT usage but places responsibilities at the correct layer imo
```

From what I can tell, this isn't too hard to deal with in NJT but it does make CacheTensor even dumber and straightforward (in a good way imo)

soulitzer replied:
Is the concern is that CacheTensor in its current state isn't constructed like one would expect a tensor named CacheTensor? e.g. in this model one would expect CacheTensor to be first and foremost a tensor, but has additional fields stored in a dict. Given this, I think the concern about its construction makes sense.

Could I alternatively propose that we changed the name of the tensor to `SimpleTensorDict`? In that model this tensor is more like a wrapper over a dict, in which every field is seen uniformly. In this world, the current way of construction should be most natural.

soulitzer replied:
Thinking out side of the NJT context, perhaps a reason to favor `SimpleTensorDict` over `CacheTensor` is that
CacheTensor doesn't actually benefit much from being first and foremost a tensor, because it immediately unwraps upon any operators call.  A container-like name of  simple-tensor-dict makes it clearer that this is more of a container object than a tensor you are meant to operate on.

jbschlosser replied:
> Is the concern is that CacheTensor in its current state isn't constructed like one would expect a tensor named CacheTensor? e.g. in this model one would expect CacheTensor to be first and foremost a tensor, but has additional fields stored in a dict. Given this, I think the concern about its construction makes sense.

Yep you nailed it here. Trying to have the impl match the mental model of such a class (for reasoning / maintenance purposes mainly - much easier to verify things with clear semantics).

> Could I alternatively propose that we changed the name of the tensor to SimpleTensorDict? In that model this tensor is more like a wrapper over a dict, in which every field is seen uniformly. In this world, the current way of construction should be most natural.
>
> Thinking out side of the NJT context, perhaps a reason to favor SimpleTensorDict over CacheTensor is that
CacheTensor doesn't actually benefit much from being first and foremost a tensor, because it immediately unwraps upon any operators call. A container-like name of simple-tensor-dict makes it clearer that this is more of a container object than a tensor you are meant to operate on.

Okay I generally like where this is going- if you really want to act on a dictionary cache, use an object that fits that model. How would the torch_dispatch for `SimpleTensorDict` work - would it still unwrap for a specific key?

soulitzer replied:
That is a good point - It is true that the current behavior is not symmetric wrt all the field... maybe what I really want to happen is to ban all operations.

soulitzer replied:
> maybe what I really want to happen is to ban all operations.

This should be fine except for the purpose of compile, I need to support 'detach'. WDYT?

jbschlosser replied:
> This should be fine except for the purpose of compile, I need to support 'detach'. WDYT?

Sure, so it's only purpose is to store tensors into a dict in a way that matches the "traceable wrapper subclass" interface? Does `detach()` detach all the tensors in the dict?

soulitzer replied:
> Sure, so it's only purpose is to store tensors into a dict in a way that matches the "traceable wrapper subclass" interface?

Exactly

>  Does detach() detach all the tensors in the dict?

I think we want to return a fresh NestedTensor object, but I'm not completely sure if I want to detach everything inside.

What tips me slightly toward not detaching overall is because generally for NT's size comparison logic, we'd prefer if the instance of the offsets, etc is the same. (I'm not sure it actually errors either when I do detach, but didn't  investigate further)

For other things it seems like it shouldn't matter:

- From the perspective of compile it seems okay not to detach because we already have the precedence of NJT not detaching offsets when the outer NJT is detached.

- From the perspective of autograd it shouldn't matter either here since detach is a no-op underneath autograd.

One might argue that one would expect a dict-container-like tensor replicate the outer commands into all the inner commands, which I don't disagree with, but at the same time, I don't really consider detach() truly public API either 
because I don't expect the user to be randomly calling into detach when no other op is supported.

jbschlosser commented on torch/nested/_internal/cached_tensor.py:
```diff
@@ -0,0 +1,131 @@
+        return out
+
+    def __repr__(self) -> str:  # type: ignore[override]
+        return f"CachedTensor({repr(self.metadata[self.source_field])})"
```
I'll suggest `CacheTensor` as a slightly tweaked name since it's a tensor that contains additional cache info.

soulitzer replied:
Sounds good!

jbschlosser reviewed (commented):
Nice work! I think this most recent iteration is much simpler and closer to be being approvable. Just got one more thing bugging me (open to being convinced otherwise) and I think the name should be tweaked a bit and then it's good to go imo.

jbschlosser commented on test/test_nestedtensor.py:
```diff
@@ -8654,6 +8654,111 @@ def test_with_factor(self):
+from torch.nested._internal.cached_tensor import CachedTensor
+
+
+class TestCachedTensor(torch.testing._internal.common_utils.TestCase):
```
bumping this - should probably still move

soulitzer replied:
Done!

jbschlosser commented on torch/nested/_internal/cached_tensor.py:
```diff
@@ -0,0 +1,131 @@
+            "source_field": self.source_field,
+            "all_fields": list(self.metadata.keys()),
+        }
+        return [k for k, v in self.metadata.items() if v is not None], ctx
```
note that the order is non-deterministic here due to the use of `items()`. Is this a problem? @bdhirsh

soulitzer replied:
Is this still the case in Python versions >=  3.7?

https://stackoverflow.com/questions/14956313/why-is-dictionary-ordering-non-deterministic

jbschlosser replied:
hm guess not :) so we're good as long as we don't support older python versions than that

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"


~~A generic CachedTensor tensor wrapper subclass that is not aware of NJTs, nested ints, etc. Users of CachedTensor customize how to use it via the `source_fields`, `extra_fields`, and `target_field`:~~
- ~~`source_fields` are fields like `offsets` and `lengths` that are complete and transitions between them are information preserving.~~
- ~~`extra_fields` on the other hand are not complete. A transition from source_field to extra_field is information losing. You cannot reconstruct a source_field with only an extra_fields.~~
- ~~A CachedTensor must contain at least one source_field.~~
- ~~`target_field` (optional) determines, the metadata of the outer tensor. If no `target_field` is provided, the metadata is determined by that of the first source field.``~~

Relevant block comment from PR:
```
    # Tensor subclass wrapping a dict of tensors, whose shape, dtype, device, etc.
    # is determined by a "source" tensor in the dict (specified by the user
    # during construction).
    #
    # This class is not super useful on its own because by default, performing any
    # operations on it will be as if you first unwrapped the CachedTensor and then
    # performed the operation on the plain source tensor (a plain tensor is also returned).
    # To leverage the extra metadata, you must register an op to perform the special logic
    # you want via register_cached_tensor_func.
    #
    # When used this way (1) it is a convenient way to keep around metadata
    # related to a tensor, without having to laboriously thread those extra metadata
    # around, e.g. through aten signatures. (2) allows one to trigger custom
    # __torch_dispatch__ logic to construct tensor subclasses (which can be tricky to do
    # otherwise because the subclass constructor op itself usually does not take the
    # subclass itself as input!). See NestedTensor for an example.
    
 ```



 

[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"


~~A generic CachedTensor tensor wrapper subclass that is not aware of NJTs, nested ints, etc. Users of CachedTensor customize how to use it via the `source_fields`, `extra_fields`, and `target_field`:~~
- ~~`source_fields` are fields like `offsets` and `lengths` that are complete and transitions between them are information preserving.~~
- ~~`extra_fields` on the other hand are not complete. A transition from source_field to extra_field is information losing. You cannot reconstruct a source_field with only an extra_fields.~~
- ~~A CachedTensor must contain at least one source_field.~~
- ~~`target_field` (optional) determines, the metadata of the outer tensor. If no `target_field` is provided, the metadata is determined by that of the first source field.``~~

Relevant block comment from PR:
```
    # Tensor subclass wrapping a dict of tensors, whose shape, dtype, device, etc.
    # is determined by a "source" tensor in the dict (specified by the user
    # during construction).
    #
    # This class is not super useful on its own because by default, performing any
    # operations on it will be as if you first unwrapped the CachedTensor and then
    # performed the operation on the plain source tensor (a plain tensor is also returned).
    # To leverage the extra metadata, you must register an op to perform the special logic
    # you want via register_cached_tensor_func.
    #
    # When used this way (1) it is a convenient way to keep around metadata
    # related to a tensor, without having to laboriously thread those extra metadata
    # around, e.g. through aten signatures. (2) allows one to trigger custom
    # __torch_dispatch__ logic to construct tensor subclasses (which can be tricky to do
    # otherwise because the subclass constructor op itself usually does not take the
    # subclass itself as input!). See NestedTensor for an example.
    
 ```



 

[ghstack-poisoned]

soulitzer commented on torch/nested/_internal/cached_tensor.py:
```diff
@@ -0,0 +1,125 @@
+
+        if op is torch.ops.aten.detach.default:
+            # detach is needed for torch.compile
+            return args[0]
```
Reminder to rewrap to fresh tensor.

bdhirsh commented on torch/nested/_internal/cached_tensor.py:
```diff
@@ -0,0 +1,125 @@
+    def __repr__(self) -> str:  # type: ignore[override]
+        return f"CachedTensor({repr(self.metadata[self.source_field])})"
+
+    def __getattr__(self, name: str) -> Optional[torch.Tensor]:
```
Is this custom getattr mainly to make sure that `cache_tensor.key` secretly maps to `cache_tensor.metadata.key`, so the source information of inner tensors is accurate?

That makes sense, although I'm not sure how good dynamo's handling of custom getattr's is. Do you ever expect attribute accesses on a `CachedTensor` to happen from user python code, that dynamo needs to inline?

soulitzer replied:
> Is this custom getattr mainly to make sure that cache_tensor.key secretly maps to cache_tensor.metadata.key, so the source information of inner tensors is accurate?

Yeah, this was mainly to get `__tensor_flatten__` to work

> That makes sense, although I'm not sure how good dynamo's handling of custom getattr's is. Do you ever expect attribute accesses on a CachedTensor to happen from user python code, that dynamo needs to inline?

Nah, today I expect CachedTensor to be purely private.

bdhirsh commented on torch/nested/_internal/cached_tensor.py:
```diff
@@ -0,0 +1,125 @@
+    def __tensor_unflatten__(
+        inner_tensors: Dict, meta: Dict, outer_size: Any, outer_stride: Any
+    ) -> "CachedTensor":
+        inner_tensors = {k: inner_tensors.get(k) for k in meta["all_fields"]}
```
aren't `inner_tensors.keys()` and `meta["all_fields"]` identical / redundant? Or is it possible for them to be different?

soulitzer replied:
Inner tensors keys are incomplete because they only contain the k's that are non-None. (I'm not sure if we can have inner tensors that are None.)

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"





 

[ghstack-poisoned]

soulitzer committed:
Update on "Add CachedTensor tensor wrapper subclass"





 

[ghstack-poisoned]

soulitzer committed:
Update on "Add DictTensor tensor wrapper subclass"





 

[ghstack-poisoned]

soulitzer commented:
Updates:
- Renamed from CachedTensor to DictTensor (see this long discussion for context https://github.com/pytorch/pytorch/pull/141841#discussion_r1887678558)
- Moved the test from test_nestedtensor to a separate file (test_dict_tensor)
- The shape of the tensor is (0,) to avoid trying to copy the metadata of some arbitrary inner tensor.

soulitzer committed:
Update on "Add DictTensor tensor wrapper subclass"





 

[ghstack-poisoned]

soulitzer committed:
Update on "Add DictTensor tensor wrapper subclass"





 

[ghstack-poisoned]

soulitzer committed:
Update on "Add DictTensor tensor wrapper subclass"





 

[ghstack-poisoned]

soulitzer committed:
Update on "Add DictTensor tensor wrapper subclass"





 

[ghstack-poisoned]

soulitzer committed:
Update on "Add DictTensor tensor wrapper subclass"





 

[ghstack-poisoned]

soulitzer committed:
Update on "Add DictTensor tensor wrapper subclass"





 

[ghstack-poisoned]

Skylion007 commented on torch/nested/_internal/dict_tensor.py:
```diff
@@ -0,0 +1,117 @@
+from contextlib import contextmanager
+from typing import *  # noqa: F403
```
I'm surprised we allow star imports.

Skylion007 commented on torch/nested/_internal/dict_tensor.py:
```diff
@@ -0,0 +1,117 @@
+# redirect calls to that op to your own implementations.
+# For NestedTensor we rely on this behavior for factory functions.
+def register_dict_tensor_func(aten_op: Any) -> Callable:
+    def wrapper(func: Callable) -> Callable:
```
This erases the return type and params of the callable, you either want to use ParamSpec or a TypeVar
